### ISSUE TITLE
Support 'why-run' mode

### ISSUE THREAD

**Issue by miketheman** [DataDog/chef-datadog#118]

**Comment by olivielpeau** [comment]
Tested on Chef 12.2.1, the why-run execution seems to work fine (no changes to the state and no event/metrics emitted).

The only case I've found that doesn't work is the `dd-handler` recipe when the `chef-handler-datadog` gem is not installed. The run fails when `'chef/handler/datadog'` is required [here](https://github.com/DataDog/chef-datadog/blob/7e81689ab6c972337409a4707ff46ab44bbe785e/recipes/dd-handler.rb#L29), with the following error:

```
  ================================================================================
  Recipe Compile Error in /home/vagrant/cookbooks/datadog/recipes/dd-handler.rb
  ================================================================================

  LoadError
  ---------
  cannot load such file -- chef/handler/datadog

  Cookbook Trace:
  ---------------
    /home/vagrant/cookbooks/datadog/recipes/dd-handler.rb:29:in `from_file'
```

(obviously because in why-run mode the `chef_gem` resource doesn't actually install the gem that's required)

But from what I understand this case is unavoidable.

**Comment by miketheman** [comment]
@olivielpeau Thanks for the investigation. Regarding the final case, this was [reported a while back](https://tickets.opscode.com/browse/CHEF-4391), and claimed to have been resolved. Might be worthwhile looking into it a bit further?

**Comment by olivielpeau** [comment]
@miketheman I think the issue you're mentioning is that the `chef_gem` resource itself breaks why-run mode, whereas here it's the subsequent `require` on `chef/handler/datadog` that's making the run fail when the file doesn't exist.

I'll look more into this issue (in particular I'll look at how chef_handler loads the handler's class and if that can help us solve this).

**Comment by miketheman** [comment]
@olivielpeau You're right - I linked the wrong issue.

Thinking about this further, there might be a better way to wrap the require that behaves differently based on whether or not we're in why-run mode, by inspecting the value of `Chef::Config[:why_run]`.

Something maybe like:
```ruby
if Chef::Config[:why_run]
  Chef::Log.warn('Running in why-run mode, skipping dd-handler')
  return
end
```

at the top of the recipe.

**Comment by olivielpeau** [comment]
@miketheman I like the idea of skipping `dd-handler` in why-run mode as I can't see any clean workaround to execute it in this mode.

FYI the `1.2.0` release of chef_handler allows using the handler without having to `require` it beforehand because it requires the handler's file on its own (see https://github.com/chef-cookbooks/chef_handler/blob/v1.2.0/providers/default.rb#L41), which would solve our issue here. The thing is that this version is only compatible with Chef >= 12 so we can't use it, but we should keep that in mind for the future.

Apart from that I think why-run mode is fully supported.

**Comment by miketheman** [comment]
Closed via #231
